### PR TITLE
ARROW-16788: [C++] Remove hardening flags gRPC doesn't support

### PR DESCRIPTION
**Pull request by lidavidm**
gRPC doesn't support building with certain flags that distros insert

### CODE REVIEW

**Comment by lidavidm** [comment]
@github-actions crossbow submit github-ubuntu-focal-amd64 travis-ubuntu-focal-arm64

**Comment by lidavidm** [comment]
@github-actions crossbow submit *focal*

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -3657,8 +3657,21 @@ macro(build_grpc)
...
   string(REPLACE ";" "|" GRPC_PREFIX_PATH_ALT_SEP "${GRPC_CMAKE_PREFIX}")
 
+  # Negate warnings that gRPC cannot build under
+  # See https://github.com/grpc/grpc/issues/29417
+  set(GRPC_C_FLAGS
+      "${EP_C_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option")
+  set(GRPC_CXX_FLAGS
+      "${EP_CXX_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option")
+  string(REPLACE "-Wformat-security" "" GRPC_C_FLAGS "${GRPC_C_FLAGS}")
+  string(REPLACE "-Wformat-security" "" GRPC_CXX_FLAGS "${GRPC_CXX_FLAGS}")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:3667]: Do we need them? I think that `-Wno-format-security` wins with `-Wformat-security -Wno-format-security`. (`-Wformat-security` is ignored.)

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -3657,8 +3657,21 @@ macro(build_grpc)
   # Yuck, see https://stackoverflow.com/a/45433229/776560
   string(REPLACE ";" "|" GRPC_PREFIX_PATH_ALT_SEP "${GRPC_CMAKE_PREFIX}")
 
+  # Negate warnings that gRPC cannot build under
+  # See https://github.com/grpc/grpc/issues/29417
+  set(GRPC_C_FLAGS
+      "${EP_C_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option")
+  set(GRPC_CXX_FLAGS
+      "${EP_CXX_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:3665]: Could you add these flags in `if(NOT MSVC) ... endif()`?

**Comment by lidavidm** [comment]
@github-actions crossbow submit ubuntu-focal-*

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -3657,8 +3657,22 @@ macro(build_grpc)
...
 
+  if(NOT MSVC)
+    # Negate warnings that gRPC cannot build under
+    # See https://github.com/grpc/grpc/issues/29417
+    set(GRPC_C_FLAGS
+        "${EP_C_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option")
+    set(GRPC_CXX_FLAGS
+        "${EP_CXX_FLAGS} -Wno-attributes -Wno-format-security -Wno-unknown-warning-option"
+    )
+  endif()
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:3668]: We need to initialize `GRPC_C*_FLAGS` for MSVC.

e.g.:

```cmake
set(GRPC_C_FLAGS "${EP_C_FLAGS}")
set(GRPC_CXX_FLAGS "${EP_CXX_FLAGS}")
if(NOT MSVC)
  set(GRPC_C_FLAGS "${GRPC_C_FLAGS} ...")
  set(GRPC_CXX_FLAGS "${GRPC_CXX_FLAGS} ...")
endif()
```

Or

```cmake
set(GRPC_CMAKE_ARGS ...)
if(NOT MSVC)
  set(GRPC_C_FLAGS "${EP_C_FLAGS} ...")
  set(GRPC_CXX_FLAGS "${EP_CXX_FLAGS} ...")
  list(APPEND GRPC_CMAKE_ARGS "-DCMAKE_C_FLAGS=${GRPC_C_FLAGS}" ...)
endif()
```

Comment (lidavidm) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:3668]: D'oh…thanks for catching that

**Review by kou — APPROVED** [review]
+1

**Pull request merged**